Kubelet: Use RepoDigest for ImageID when available

DirectXMan12 opened the pull request:
**Release note**:

``` release-note
Use manifest digest (as `docker-pullable://`) as ImageID when available (exposes a canonical, pullable image ID for containers).
```

Previously, we used the docker config digest (also called "image ID"
by Docker) for the value of the `ImageID` field in the container status.
This was not particularly useful, since the config manifest is not
what's used to identify the image in a registry, which uses the manifest
digest instead.  Docker 1.12+ always populates the RepoDigests field
with the manifest digests, and Docker 1.10 and 1.11 populate it when
images are pulled by digest.

This commit changes `ImageID` to point to the the manifest digest when
available, using the prefix `docker-pullable://` (instead of
`docker://`)

Related to #32159

<!-- Reviewable:start -->

---

This change is [<img src="https://reviewable.kubernetes.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.kubernetes.io/reviews/kubernetes/kubernetes/33014)

<!-- Reviewable:end -->

DirectXMan12 commented:
This is an implementation of the short-term solution I discussed in #32159.

cc @kubernetes/sig-node @derekwaynecarr @ncdc

DirectXMan12 commented:
cc @dchen1107

derekwaynecarr commented:
/cc @deads2k

ncdc commented on pkg/kubelet/dockertools/docker.go:
```diff
@@ -153,6 +154,14 @@ func filterHTTPError(err error, image string) error {
 func matchImageTagOrSHA(inspected dockertypes.ImageInspect, image string) bool {
+	// NB: the code below this check checks for validity of `image` as a pull spec.
+	// However, it can be useful to also inspect images by ID (config digest, e.g.
+	// when determining the manifest digestof the image in use by a container).
```
s/digestof/digest of/

ncdc commented on pkg/kubelet/dockertools/docker.go:
```diff
@@ -202,6 +211,8 @@ func matchImageTagOrSHA(inspected dockertypes.ImageInspect, image string) bool {
 		if digest.Digest().Algorithm().String() == id.Algorithm().String() && digest.Digest().Hex() == id.Hex() {
 			return true
+		} else {
+			glog.Infof("DIGEST: %v != %v", digest.Digest().Hex(), id.Hex())
```
This probably needs to be V(4), and it could be a bit more verbose :-)

DirectXMan12 replied:
Whoops, I think that suck in from debugging.

ncdc commented on pkg/kubelet/dockertools/docker.go:
```diff
@@ -153,6 +154,14 @@ func filterHTTPError(err error, image string) error {
 // Check if the inspected image matches what we are looking for
 func matchImageTagOrSHA(inspected dockertypes.ImageInspect, image string) bool {
+	// NB: the code below this check checks for validity of `image` as a pull spec.
+	// However, it can be useful to also inspect images by ID (config digest, e.g.
```
I'm not sure I'd ever expect this to happen. You would have to pull the image by tag or digest to a node, then run a pod whose container's image is the value for `config.digest`, and that pod would have to be scheduled to the same node. Right?

DirectXMan12 replied:
I'm not sure I'm understanding you here, but when you do container-inspect, you get a field called `ImageID`, which we currently use to populate the `ImageID`.  In order to turn that into a manifest digest, you have to do an image-inspect on that config digest to get the information for the image (which includes the `RepoDigests` field).

ncdc replied:
The purpose of this function is to determine if the `image` parameter matches the `inspected` image, either by tag or digest. Because there are 3 ways a user can reference an image (tag, digest, id aka config.digest), we have to make sure that the inspected image does not match id/config.digest.

Because `inspected.ID` is the value of `config.digest`, the only way for it to equal `image` is if the user specified the `config.digest` for `containers[i].image` in the pod spec. But you can't pull by `config.digest`, which means that the only way you can successfully run a container whose image is `config.digest` is if that image already exists on the node. So you would have to:

1. Run a pod whose image is e.g. `foo/bar:latest`
1. Find out what that image's `config.digest` is
1. Run another pod, specifying the `config.digest` value for the image, **and** make sure the pod is scheduled to the same node from step 1

DirectXMan12 replied:
So, the only way to get into the docker `InspectImage` API is through the `InspectImage` method on `DockerInterface`, which calls this.  So the alternative is to add an `InspectImageUnchecked` to `DockerInterface` which doesn't verify the inspected info after pulling (or add a `check` argument to the `InspectImage` method).

I see what you mean here, though.  The intent of the method was not entirely clear based on the godoc above :-/.

DirectXMan12 replied:
ok, switched over to adding a `checked` argument to `InspectImage` which optionally skips the check.

yujuhong replied:
I don't like the `checked` arg added to the function. If the verification is not always needed, why not just move it up to somewhere else? E.g., `IsImagePresent` might be a good candidate.

ncdc commented on pkg/kubelet/dockertools/docker.go:
```diff
@@ -45,6 +45,7 @@ import (
 const (
 	PodInfraContainerName = leaky.PodInfraContainerName
 	DockerPrefix          = "docker://"
+	DockerPullablePrefix  = "docker-pullable://"
```
Is there a reason we need a different prefix?

DirectXMan12 replied:
It lets us tell which value (config digest vs manifest digest/spec) we're getting at a glance.  Otherwise, there's not an easy way to tell, except by trying to parse and/or inspect the value

ncdc commented on pkg/kubelet/dockertools/docker_manager.go:
```diff
@@ -397,11 +397,26 @@ func (dm *DockerManager) inspectContainer(id string, podName, podNamespace strin
+	// default to the image ID, but try and inspect for the RepoDigests
+	imageID := DockerPrefix + iResult.Image
+	imgInspectResult, err := dm.client.InspectImage(iResult.Image)
+	if err != nil {
```
Do you think this could be cleaner as a switch block?

DirectXMan12 replied:
hmm... like

```go
switch {
case err != nil:
  utilruntime.HandleError(fmt.Errorf("unable to inspect docker image %q while inspecting docker container %q: %v", containerName, iResult.Image, err))
case len(imgInspectResult.RepoDigests) > 1:
  glog.V(4).Infof("Container %q had more than one associated RepoDigest (%v), only using the first", containerName, imgInspectResult.RepoDigests)
  fallthrough
case len(imgInspectResult.RepoDigests) > 0:
  imageID = xyz
```
Maybe?  I could also just move the logging if statement out a bit (it doesn't need to be nested like it is).  I'm not convinced that a switch is better than that

ncdc replied:
Yeah, something like that. I was trying to get rid of the nesting if possible.

derekwaynecarr commented:
FYI @smarterclayton

simon3z commented:
Thanks for reusing the `imageID` field for this, it works well for us (ManageIQ/CloudForms).

Although last time I heard from @deads2k I thought we weren't going to change the prefix (`docker://` vs `docker-pullable://`). Having this information is even better for us so we can actually tell whether we're able to crosslink this with running Pods or not.

simon3z commented:
@enoodle @zeari please note that the prefix could change to `docker-pullable://` (for your current work on image crosslinking).

derekwaynecarr commented:
@dchen1107 -- are you able to review this?  @DirectXMan12 -- can you make the red x's go away?

dchen1107 reviewed (commented):
I am ok with the change in general, especially it only expose the image ID without changing any behavior for now.

yujuhong commented:
> @dchen1107 in the long run, we want to have two separate fields, but we were thinking that we could start here, since we've never explicitly guaranteed the meaning of the field (and you'll be able to distinguish between the old and new meanings using the prefix change).

This PR effecitvely changes the meaning of the `ImageID` field in the container status. I'm not sure if that's acceptable for the users. but assume it is (meaning no users want the config.digest), why would we ever want to change to "two separate fields" in the long run?

vishh commented:
ImageID was only meant for debugging. Will this change be necessary if the `Image` in the PodSpec is a digest?

dchen1107 commented:
@yujuhong This is my concern too. I understand the motivation why @DirectXMan12 want to start from here. But once we build the automated system and logic based on top of this, especially relying on parsing the certain string, it is difficult to change later. My concern might not valid in this case. 

@vishh I think there is a need to expose the right / canonical image id, at least at the node level. Given the current situation, before settling down the generic structure to represent different image types, I don't think we can simply introducing digest to PodSpec. What @DirectXMan12 proposed here is a simple and less intrusive way to expose the necessary information, and this change enables building an out-of-band image / package management system at the cluster level for them. What they learnt from that system eventually can benefit OSS kubernetes to define a generic structure for other image types, and build a generic package / image management system at cluster level.

yifan-gu commented:
cc @euank @jonboulle

dchen1107 commented:
@DirectXMan12 Can you fix the failing tests?

yujuhong commented:
Quoting myself so the https://github.com/kubernetes/kubernetes/pull/33014#discussion_r80579397 doesn't get lost:
"I don't like the `checked` arg added to the function. If the verification is not always needed, why not just move it up to somewhere else? E.g., `IsImagePresent` might be a good candidate."

ncdc commented:
I agree w/@yujuhong about disliking the `checked` argument

DirectXMan12 commented:
> E.g., `IsImagePresent` might be a good candidate

There's also `RemoveImage` in the docker_manager, which should probably have the same logic (but feel free to correct me -- I'm not as familiar with the node code, currently).  Thus, we'd probably want an `InspectImageUnchecked` or something like that, which doesn't seem much better than the `checked` argument (perhaps slightly in aesthetics, since Go doesn't have named arguments). @ncdc @yujuhong  WDYT?

DirectXMan12 commented:
@k8s-bot gci gke e2e test this

it appears to have just blown up running `kube-up`

ncdc commented:
`InspectImage` is currently called in 3 places:

1. `DockerManager.RemoveImage`
  1. As best I can tell, this is only used by image garbage collection, and it ultimately uses the ID of the image that was returned from the "list images" call to the docker daemon, which is the `config.digest` value, not the pull-by-digest value. 
1. `DockerManager.isImageRoot`
  1. The "image" that this function inspects is the value from the `api.Container`, so it will be whatever the user specified, i.e. a tag or a pull-by-digest value.
1. `dockerPuller.IsImagePresent`
  1. It looks like this the image from the `api.Container`, so it should be the user-specified tag/pull-by-digest value

The original PR that added `matchImageTagOrSHA` wanted to ensure that a pod spec that had an image named `foo/bar:2208` (a tag of 2208) didn't accidentally end up running an image in the same repository, yet with a different tag, whose ID (`config.digest`) was a prefix match for the requested tag.

One thing we might consider is splitting InspectImage into InspectImageByID (requires an exact match on the `config.digest` value) and InspectImageByRef (requires an exact match on either tag or digest). Anything user-specified has to go through InspectImageByRef, and anything that operates on image identifiers returned by the system (e.g. the image field in a container inspection result) has to go through InspectImageByID. Thoughts?

DirectXMan12 committed:
Add method to inspect Docker images by ID

Previously, the `InspectImage` method of the Docker interface expected a
"pullable" image ref (name, tag, or manifest digest).  If you tried to
inspect an image by its ID (config digest), the inspect would fail to
validate the image against the input identifier.  This commit changes
the original method to be named `InspectImageByRef`, and introduces a
new method called `InspectImageByID` which validates that the input
identifier was an image ID.

DirectXMan12 commented:
@ncdc updated according to your research, PTAL

DirectXMan12 committed:
Kubelet: Use RepoDigest for ImageID when available

Previously, we used the docker config digest (also called "image ID"
by Docker) for the value of the `ImageID` field in the container status.
This was not particularly useful, since the config manifest is not
what's used to identify the image in a registry, which uses the manifest
digest instead.  Docker 1.12+ always populates the RepoDigests field
with the manifest digests, and Docker 1.10 and 1.11 populate it when
images are pulled by digest.

This commit changes `ImageID` to point to the the manifest digest when
available, using the prefix `docker-pullable://` (instead of
`docker://`)

ncdc commented on pkg/kubelet/dockertools/docker_manager.go:
```diff
@@ -397,11 +397,27 @@ func (dm *DockerManager) inspectContainer(id string, podName, podNamespace strin
+		}
+
+		if len(imgInspectResult.RepoDigests) > 0 {
+
```
delete blank line

ncdc commented on pkg/kubelet/dockertools/docker_test.go:
```diff
@@ -315,6 +315,95 @@ func TestMatchImageTagOrSHA(t *testing.T) {
+			Output: false,
+		},
+		{
+			// RepoDigest match is allowed
```
Other than docker.io how is this test case different from the previous one?

DirectXMan12 replied:
it's not really for this method.  I forgot to remove it.

ncdc commented:
@DirectXMan12 I'd appreciate it if you could sanity check my research and confirm my findings

yujuhong commented:
@DirectXMan12 I'd appreciate if you can add a node e2e test.

DirectXMan12 commented:
> @DirectXMan12 I'd appreciate it if you could sanity check my research and confirm my findings

Yep, I checked while I was writing the new code ;-).  I think your findings LGTM.

ncdc commented:
FWIW, the distinction between inspecting by ref vs id can probably go away once our minimum supported docker version is whatever version contains the fix for https://github.com/docker/docker/issues/22112 (stop allowing references like foo/bar:1234 from prefix matching image id)

euank commented:
> in the long run, we want to have two separate fields, but we were thinking that we could start here, since we've never explicitly guaranteed the meaning of the field (and you'll be able to distinguish between the old and new meanings using the prefix change).

This is not a backwards compatible change. Even if we didn't promise the semantics, they existed as de-facto semantics (and could be derived as explicit semantics by reading the code). Quoting from [doc](https://github.com/kubernetes/kubernetes/blob/master/docs/devel/api_changes.md#on-compatibility):

> A new representation, even in a new API version, that is more expressive than an old one breaks backward compatibility, since clients that only understood the old representation would not be aware of the new representation nor its semantics.

That's also true for migrating `docker-pullable` away into its own field in the future. I see this change as a backwards incompatible change with plans for a future additional one.

If we're going to spit it out later anyways, why not do it now? Is it to try and avoid the overhead of adding a field, or is there a better reason?

@bgrant0607 or @smarterclayton, opinions WRT api-compatibility of changing `ImageID` from always having the form `docker://` to having a possible form of `docker-pullable://` or `docker://`?

DirectXMan12 commented:
> @DirectXMan12 I'd appreciate if you can add a node e2e test.

Yep, will do.  Just having some issues getting the e2e tests running properly in my test setup :-/.

smarterclayton commented:
I think this one sits in the line between intent and compatibility.  We fix bugs in broken use cases (i.e. if it's not useful to the cluster consumer, we shouldn't be setting it).  However, I would agree in general that we try to be sticklers for preserving the semantics of a field, and this is changing the semantics of a field.  

We in the past have chosen one of three options:

1. Fix the field for everyone to be consistent (in this case, because the value changes from use to use it's harder to justify)
2. Add a flag that triggers the new behavior
3. Add a new field alongside the existing field that contains the improved information.

This case sounds like either 2 or 3, because 1 is difficult to make correct.  The problem with 3 is that we'd be adding a new field that doesn't take into account the changes we desire to make in the future about image information.  I'm slightly concerned about designing for that change here.

That probably leaves #2 as the safest option, even though I would prefer 3.

smarterclayton commented:
I would still argue the field as is useless - if we were to redefine it to mean something new, it's no worse than the current behavior.  I think this is a call Brian probably needs to help weigh in on.

ncdc commented:
I was trying to find a way to describe the current state of the field, and "useless" is a pretty good one 😄, at least for docker image IDs.

derekwaynecarr commented:
@euank - i thought we left the sig-node call last week agreeing that the change was fine as the field had no definition today asserted in the API.  /cc @dchen1107

derekwaynecarr commented:
To clarify, I think a flag is overkill, as I don't know why you would NOT want to always have this flag enabled.

dchen1107 commented:
I agreed that that field representing in today's way is useless.

DirectXMan12 commented:
Added an e2e, should be all set.

euank commented:
@derekwaynecarr My understanding was that we'd add a new one. Sorry for my misunderstanding and getting to this conversation a bit late; my bad!

My concern here is more trying to "keep us honest" to our own standards than practical, and if we want to assert that we don't need to follow the rules for this sort of mostly-useless-field case I'm fine with being overruled.

derekwaynecarr commented:
@k8s-bot kubemark e2e test this

ncdc commented:
LGTM

derekwaynecarr commented:
@k8s-bot kubemark e2e test this

derekwaynecarr commented:
@dchen1107 -- looks like all the check marks are green... ptal

bgrant0607 commented:
@euank ImageID will always contain `docker-pullable://...` after this change?

The reason the value in that field contains a prefix is so that we could add other types of image IDs, but we expected to change it upon request only, such as for other kinds of images, like OCI.

@smarterclayton @euank @dchen1107 Why is the current field useless? Feel free to point me at an earlier comment.

euank commented:
@bgrant0607 The current image id it presents, `docker://<value>`, displays a node-local value which cannot be used to reference the image either on external registries nor other nodes.

This change is about making it so it can whenever possible display a value that is not node-local, but rather a uniquely identifying and reference-able in external repositories. The new value will not always be present iiuc, and the `docker://` prefix could still be visible (e.g. on nodes with old docker clients or in the case of other errors).

Since the original spirit of including the prefix was for users to parse and switch on it with the expectation of more, I withdraw my concern.
Thanks for the info, this now LGTM. I also realize that this field already is overloaded to include the `rkt://` prefix in that case.

My bad for my bad understanding drawing this out, sorry!

yujuhong commented:
> @euank ImageID will always contain docker-pullable://... after this change?

After this change, if an image digest is available, kubele will report `docker-pullable://`. Otherwise, it'd report `docker://`. The digest may not be available if you pulled by name:tag originally.

derekwaynecarr commented:
@bgrant0607 -- it sounds like there are no more remaining concerns?  Objections to merging?

bgrant0607 commented:
@derekwaynecarr I discussed the change with @yujuhong and it's ok with me from an API-compatibility point of view. She'll make a final pass over the PR.

yujuhong commented:
LGTM.

wojtek-t commented:
This PR broke all kubemark suites. I'm reverting it.
